golang/cosmos/x/swingset/keeper/store_snapshotter.go

```diff
@@ -0,0 +1,601 @@
+const UntrustedExportDataArtifactName = "UNTRUSTED-EXPORT-DATA"
+
+const exportDataFilename = "export-data.jsonl"
+const untrustedExportDataFilename = "untrusted-export-data.jsonl"
+const exportedFilesMode = 0644
+
+var disallowedArtifactNameChar = regexp.MustCompile(`[^-_.a-zA-Z0-9]`)
+
+// sanitizeArtifactName searches a string for all characters
+// other than ASCII alphanumerics, hyphens, underscores, and dots,
+// and replaces each of them with a hyphen.
+func sanitizeArtifactName(name string) string {
+	return disallowedArtifactNameChar.ReplaceAllString(name, "-")
+}
+
+type activeSnapshot struct {
+	// Whether the operation in progress is a restore
+	isRestore bool
+	// The block height of the snapshot in progress
+	height uint64
+	// The logger for this snapshot
+	logger log.Logger
+	// Use to synchronize the commit boundary
+	startedResult chan error
+	// Internal flag indicating whether the snapshot was retrieved
+	// Only read or written by the snapshot worker goroutine.
+	retrieved bool
+	// Closed when this snapshot is complete
+	done chan error
+}
+
+type exportManifest struct {
+	BlockHeight uint64 `json:"blockHeight,omitempty"`
+	// The filename of the export data
+	Data string `json:"data,omitempty"`
+	// The list of artifact names and their corresponding filenames
+	Artifacts [][2]string `json:"artifacts"`
+}
+
+// SwingStoreExportProvider gives access to a SwingStore snapshot export data
+// and the related artifacts.
+// The abstraction is similar to the JS side swing-store exporter abstraction,
+// but without the ability to list artifacts or random access them.
+type SwingStoreExportProvider struct {
+	// The block height of the SwingStore export
+	BlockHeight uint64
+	// A function to get the "export data" of the SwingStore export, if any
+	// errors with io.EOF if no "export data" is available for this export
+	GetExportData func() ([]*vstoragetypes.DataEntry, error)
+	// A function to retrieve each artifact in the SwingStore export
+	// errors with io.EOF once reaching the end of the artifact list
+	ReadArtifact func() (types.SwingStoreArtifact, error)
+}
+
+// SwingStoreSnapshotTaker is used to delegate operations that occur while
+// creating a snapshot. It defines the interface of the object implementing
+// these operations, and which is provided to InitiateSnapshot.
+type SwingStoreSnapshotTaker interface {
+	// Called after the SwingStore snapshot was initiated in a goroutine
+	// Must call the `retrieveSwingStoreExport` function before returning, which
+	// will in turn synchronously invoke `ProcessExportedSnapshot` once the
+	// snapshot is ready
+	TakeSnapshot(height uint64, retrieveSwingStoreExport func() error) error
+	// Called when the SwingStore snapshot has been retrieved
+	// The implementation must synchronously consume the provider, which becomes
+	// invalid after the method returns.
+	ProcessExportedSnapshot(provider SwingStoreExportProvider) error
```

Comment:
Can these doc comments be reworded to explicitly distinguish for each function:
* its purpose/effects/etc., and
* when it is expected to be invoked (and by what), and
* what requirements it must satisfy.

?